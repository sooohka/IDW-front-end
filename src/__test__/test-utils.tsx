/* eslint-disable @typescript-eslint/no-shadow */
/* eslint-disable no-param-reassign */
import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import renderer from "react-test-renderer";
import store from "../store";
import GlobalStyle from "../style/GlobalStyle";
import theme from "../style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const thunk =
  ({ dispatch, getState }: any) =>
  (next: any) =>
  (action: any) => {
    if (typeof action === "function")
      // 여기서 반환되는 action은 thunk함수임
      return action(dispatch, getState);
    return next(action);
  };

const useStore = (stateName: string, initState: any, reducer: any) => {
  const store = {
    dispatch: jest.fn((action: any) => {
      initState = reducer(initState, action);
    }),
    getState: jest.fn(() => ({ [stateName]: initState })),
  };
  const next = jest.fn();
  const makeThunk = (action: any) => thunk(store)(next)(action);
  return { store, next, makeThunk };
};

const createWithThemeProvider = (EL: ReactElement) => {
  return renderer.create(
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {EL}
    </ThemeProvider>,
  );
};

const redux = { thunk, useStore };
export { redux, customRender as render, createWithThemeProvider };

export * from "@testing-library/react";
