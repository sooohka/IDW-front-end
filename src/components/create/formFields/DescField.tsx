import { useFormikContext } from "formik";
import React from "react";
import styled from "styled-components";
import HelperText from "../../common/HelperText";

const StyledField = styled.div`
  display: flex;
  flex-direction: column;
`;

const FieldTitle = styled.span`
  display: flex;
  flex-direction: column;
  font-weight: bold;
  letter-spacing: 1px;
  font-size: ${({ theme }) => theme.fonts.label};
`;

const TextArea = styled.textarea`
  border: 3px solid;
  border-radius: 5px;
  border-color: ${({ theme }) => theme.colors.primary};
  padding: 1rem;
  font-size: ${({ theme }) => theme.fonts.strongBody};
  letter-spacing: 0.5px;
  font-weight: bold;
`;

interface IProps {
  name: keyof CreateFormValues;
}

const DescField: React.FC<IProps> = ({ name }) => {
  const { values, touched, errors, handleChange, handleBlur } =
    useFormikContext<CreateFormValues>();
  return (
    <>
      <StyledField>
        <FieldTitle>설명</FieldTitle>
        <TextArea
          tabIndex={0}
          name={name}
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.desc}
        />
      </StyledField>
      <HelperText
        hasError={Boolean(touched[name] && errors[name])}
        text={(touched[name] && errors[name]) as string}
      />
    </>
  );
};

export default DescField;
