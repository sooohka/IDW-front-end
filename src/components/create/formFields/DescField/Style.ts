import styled from "styled-components";

const Field = styled.div`
  display: flex;
  flex-direction: column;
`;

const FieldTitle = styled.span`
  display: flex;
  flex-direction: column;
  font-weight: bold;
  letter-spacing: 1px;
  font-size: ${({ theme }) => theme.fonts.label};
`;

const TextArea = styled.textarea`
  border: 3px solid;
  border-radius: 5px;
  border-color: ${({ theme }) => theme.colors.primary};
  padding: 1rem;
  font-size: ${({ theme }) => theme.fonts.strongBody};
  letter-spacing: 0.5px;
  font-weight: bold;
`;

export { Field, FieldTitle, TextArea };
