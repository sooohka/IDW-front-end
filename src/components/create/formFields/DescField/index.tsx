import React from "react";
import styled from "styled-components";
import HelperText from "../../../common/HelperText";

const StyledField = styled.div`
  display: flex;
  flex-direction: column;
`;

const FieldTitle = styled.span`
  display: flex;
  flex-direction: column;
  font-weight: bold;
  letter-spacing: 1px;
  font-size: ${({ theme }) => theme.fonts.label};
`;

const TextArea = styled.textarea`
  border: 3px solid;
  border-radius: 5px;
  border-color: ${({ theme }) => theme.colors.primary};
  padding: 1rem;
  font-size: ${({ theme }) => theme.fonts.strongBody};
  letter-spacing: 0.5px;
  font-weight: bold;
`;

interface IProps {
  name: keyof CreateFormValues;
  handleBlur: React.FocusEventHandler;
  handleChange: React.ChangeEventHandler;
  value: string;
  error?: string;
  touched?: boolean;
}

const DescField: React.FC<IProps> = ({ name, handleChange, handleBlur, value, touched, error }) => {
  return (
    <>
      <StyledField>
        <FieldTitle>설명</FieldTitle>
        <TextArea
          tabIndex={0}
          name={name}
          onChange={handleChange}
          onBlur={handleBlur}
          value={value}
        />
      </StyledField>
      <HelperText hasError={Boolean(touched && error)} text={(touched && error) as string} />
    </>
  );
};

export default DescField;
