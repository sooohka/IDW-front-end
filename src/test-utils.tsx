import { render, RenderOptions } from "@testing-library/react";
import React, { FC, ReactElement } from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import store from "./store";
import GlobalStyle from "./style/GlobalStyle";
import theme from "./style/theme";

const AllTheProviders: FC = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          <GlobalStyle />
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
