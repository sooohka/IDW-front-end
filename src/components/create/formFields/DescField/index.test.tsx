import userEvent from "@testing-library/user-event";
import { createWithThemeProvider, render, screen } from "../../../../__test__/test-utils";
import DescField from ".";

describe("components/create/formFields/DescFields", () => {
  let name: keyof CreateFormValues;
  let error: string | undefined;
  let touched: boolean | undefined;
  let handleBlur: any;
  let handleChange: any;
  let value: any;
  let EL: any;

  beforeEach(() => {
    name = "desc";
    error = undefined;
    touched = undefined;
    handleBlur = jest.fn();
    handleChange = jest.fn();
    value = "";
    EL = (
      <DescField
        error={error}
        touched={touched}
        handleBlur={handleBlur}
        handleChange={handleChange}
        value={value}
        name={name}
      />
    );
  });
  it("renders", () => {
    const component = createWithThemeProvider(EL);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("type something", () => {
    render(EL);
    const textField = screen.getByRole("textbox");
    userEvent.type(textField, "hello");
    expect(handleChange).toBeCalledTimes(5);
  });
});
