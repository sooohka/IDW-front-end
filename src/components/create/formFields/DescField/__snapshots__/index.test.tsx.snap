// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/create/formFields/DescFields renders 1`] = `
Array [
  <div
    className="sc-gsDKAQ gFZcdC"
  >
    <span
      className="sc-dkPtRN jNMDnb"
    >
      설명
    </span>
    <textarea
      className="sc-hKwDye PbQHt"
      name="desc"
      onBlur={[MockFunction]}
      onChange={[MockFunction]}
      tabIndex={0}
      value=""
    />
  </div>,
  <p
    className="sc-bdvvtL ecvaSW"
    role="note"
  />,
]
`;
